### ISSUE TITLE
Lock file does not lock

### ISSUE THREAD

**Issue by farcop** [chainguard-dev/rules_apko#13]
I used lockfile from https://github.com/chainguard-dev/rules_apko/issues/7#issuecomment-1721697367 and built my base image successfully.

Today if I try to build my project I get the error:
```
ℹ️  aarch64   | determining desired apk world
Error: failed to build layer image for "arm64": installing apk packages: installing packages: expanding busybox: fetching package "busybox": unable to get package apk at https://dl-cdn.alpinelinux.org/alpine/edge/main/aarch64/busybox-1.36.1-r8.apk: Get "https://dl-cdn.alpinelinux.org/alpine/edge/main/aarch64/busybox-1.36.1-r8.apk": failed to read "bazel-out/darwin_arm64-fastbuild/bin/images/alpine/cache_image/https%3A%2F%2Fdl-cdn.alpinelinux.org%2Falpine%2Fedge%2Fmain/aarch64/busybox-1.36.1-r8.apk" in offline cache: open bazel-out/darwin_arm64-fastbuild/bin/images/alpine/cache_image/https%3A%2F%2Fdl-cdn.alpinelinux.org%2Falpine%2Fedge%2Fmain/aarch64/busybox-1.36.1-r8.apk: no such file or directory
2023/09/20 08:37:59 error during command execution: failed to build layer image for "arm64": installing apk packages: installing packages: expanding busybox: fetching package "busybox": unable to get package apk at https://dl-cdn.alpinelinux.org/alpine/edge/main/aarch64/busybox-1.36.1-r8.apk: Get "https://dl-cdn.alpinelinux.org/alpine/edge/main/aarch64/busybox-1.36.1-r8.apk": failed to read "bazel-out/darwin_arm64-fastbuild/bin/images/alpine/cache_image/https%3A%2F%2Fdl-cdn.alpinelinux.org%2Falpine%2Fedge%2Fmain/aarch64/busybox-1.36.1-r8.apk" in offline cache: open bazel-out/darwin_arm64-fastbuild/bin/images/alpine/cache_image/https%3A%2F%2Fdl-cdn.alpinelinux.org%2Falpine%2Fedge%2Fmain/aarch64/busybox-1.36.1-r8.apk: no such file or directory
Target //images/alpine:test failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 27.318s, Critical Path: 1.24s
INFO: 361 processes: 361 internal.
FAILED: Build did NOT complete successfully
ERROR: Build failed. Not running target
```

if we look up into apko.lock.json then we will see that we already lock this package with another version:
```
      {
        "name": "busybox",
        "version": "1.36.1-r7",
        "architecture": "x86_64",
        "url": "https://dl-cdn.alpinelinux.org/alpine/edge/main/x86_64/busybox-1.36.1-r7.apk",
```

I supposed that once lock file generated and attached into workspace with translate_apko_lock then we have reproducible build at any time later.

**Comment by farcop** [comment]
It's a pity, but I still don't understand the point of using rules_apko.
The only value for me was the ability to build repeatable builds after locking apk-packages list. But I cant.

I got the script to work https://github.com/chainguard-dev/rules_apko/blob/main/resolve-all.sh  but it helps one-two days, after that build fails again.

**Comment by farcop** [comment]
It happened again. 

busybox-1.36.1-r8 -> busybox-1.36.1-r9

As soon as a new version of the package is released, the apko_image() breaks with error above and you should to re-run `apko resolve`... 
I guess this is not expected behavior.

**Comment by thesayyn** [comment]
> I guess this is not expected behavior.

We are aware of the problem and have a few things that would fix this.  However, after talking to @imjasonh, we decided that this doesn't need to be fixed right away as we'd like to get some more mileage on rules_apko before we commit more time.

**Comment by farcop** [comment]
@thesayyn I don't agree that this is a feature request. This is a bug, since the way it works now excludes the possibility of using it for its intended purpose. When build fails I have no choice, I should only use latest version of packages. This fact rules out repatable builds at any time.

**Comment by jonjohnsonjr** [comment]
Looked into this a bit and it is a very small diff to get `apko build` to use a lockfile instead of an APKINDEX.

I'll try to test those changes with rules_apko next week and get that merged.

**Comment by sfc-gh-ptabor** [comment]
@thesayyn: 
  The challenge is that `wolfi repository` is mutable and it does not keep the old versions of artifacts.
  
  But even alpine is doing breaking changes. 
  Please see: https://dl-cdn.alpinelinux.org/alpine/v3.18/community/x86_64/ (stable version 3.18): 
  
```
chromium-119.0.6045.159-r0.apk                     29-Nov-2023 04:18    109M
```

2 days ago it was `chromium-117....` . They keep removing old packages in scope the single version as well.

**Comment by sfc-gh-ptabor** [comment]
I think I understand the problem now better. There are 2 types of issues: 
  1) Fixed-version artifact in remote repository is removed (upgraded to a new version) -> this can be healed by caching layer like "Artifactory"
  2) APKINDEX get's updated to so resolution process performed by APKO evolves newer versions of packages than the ones listed in the .resolved.file. 

The 2) is more significant problem as I'm not aware of any workaround. I agree with @farcop that it makes `rules_apko` basically useless for any hermetic application. 

To solve this I would consider: 
  1.) Adding to `apko` a flag --lock_file=apko.resolved.json
      The flag would cause post-processing of APKINDEX file before passing it to go-apk, such that only artifacts present in apko.resolved.json are present in the effective APKINDEX... This way the apk resolution process can only resolve between versions that are in the APK file. 
  2.) Makes `apko_image` rule pass the `apko.resolved.json` file as a flag to the apko.

**Comment by imjasonh** [comment]
One slight clarification: normally Wolfi will not delete old APKs like Alpine does, or at least not aggressively. We may delete broken/non-functional packages from time to time, but in general we plan to keep packages around indefinitely.

Alpine's behavior does unfortunately make rules_apko quite a bit less useful than using it with Wolfi. I'm not sure what the best solution is there -- perhaps resolution should also fetch and store the apks in the source repo, so you're not reliant on what Alpine/Wolfi is currently serving? You'd still need to verify packages with the signing key they currently serve, but those don't rotate often and AFAIK aren't reaped aggressively at all during normal operation.

The APKINDEX is updated constantly, so anything that depends on the etag of the APKINDEX being served by the repos will break fairly often as new packages are added to the repository. Perhaps caching that in the source repo is a solution to that too?

**Comment by sfc-gh-ptabor** [comment]
> The APKINDEX is updated constantly, so anything that depends on the etag of the APKINDEX being served by the repos will break fairly often as new packages are added to the repository. Perhaps caching that in the source repo is a solution to that too?

The APKINDEX is huge and redundant in role in such a case to `apko.resolved.json`. I think it would need to be a trimmed APKINDEX file that contains only needed (package-version) records. 

In case of using a proxy layer like Artifactory... I'm not sure, but assume it might be Artifactory role to provide APKINDEX that contains all versions of packages that are present in the 'cached' repository -> including versions deleted upstream. 
If that's true, the `trim APKINDEX in apko` approach would work.

**Comment by sfc-gh-ptabor** [comment]
There is chain of 3 changes addressing this effort:  
 - go-apk: https://github.com/chainguard-dev/rules_apko/issues/13#:~:text=Add%20support%20for%20apk.InstallPackages%20that%20assumes%20pre%2Dresolved%20data.%20chainguard%2Ddev/go%2Dapk%23172
 - apko: https://github.com/chainguard-dev/rules_apko/issues/13#:~:text=Support%20for%20locking%20packages%20to%20the%20versions%20from%20the%20%27resolved.apko.json%27%20file.%20chainguard%2Ddev/apko%23979
 - rules_apko: https://github.com/chainguard-dev/rules_apko/pull/41
 
 @jonjohnsonjr @imjasonh - please take a look.

**Comment by sfc-gh-ptabor** [comment]
I think this one can be marked as addressed/done/completed - especially when rules_apko 1.2.0 will be released.

**Comment by imjasonh** [comment]
https://github.com/chainguard-dev/rules_apko/releases/tag/v1.2.0 🎉

**Comment by jeremyeim** [comment]
Is there a plan to publish this version to the BCR to use with bzlmod?
